python/pyspark/pandas/frame.py

```diff
@@ -7637,102 +7706,35 @@ def merge(
-                raise ValueError("Must pass left_on or left_index=True")
-            if not left_key_names and not right_key_names:
-                common = list(self.columns.intersection(right.columns))
-                if len(common) == 0:
-                    raise ValueError(
-                        "No common columns to perform merge on. Merge options: "
-                        "left_on=None, right_on=None, left_index=False, right_index=False"
-                    )
-                left_key_names = list(map(self._internal.spark_column_name_for, to_list(common)))
-                right_key_names = list(map(right._internal.spark_column_name_for, to_list(common)))
-            if len(left_key_names) != len(right_key_names):
-                raise ValueError("len(left_keys) must equal len(right_keys)")
-
-        # We should distinguish the name to avoid ambiguous column name after merging.
-        right_prefix = "__right_"
-        right_key_names = [right_prefix + right_key_name for right_key_name in right_key_names]
-
-        how = validate_how(how)
-
-        def resolve(internal: InternalFrame, side: str) -> InternalFrame:
-            rename = lambda col: "__{}_{}".format(side, col)
-            internal = internal.resolved_copy
-            sdf = internal.spark_frame
-            sdf = sdf.select(
-                *[
-                    scol_for(sdf, col).alias(rename(col))
-                    for col in sdf.columns
-                    if col not in HIDDEN_COLUMNS
-                ],
-                *HIDDEN_COLUMNS,
-            )
-            return internal.copy(
-                spark_frame=sdf,
-                index_spark_columns=[
-                    scol_for(sdf, rename(col)) for col in internal.index_spark_column_names
-                ],
-                index_fields=[
-                    field.copy(name=rename(field.name)) for field in internal.index_fields
-                ],
-                data_spark_columns=[
-                    scol_for(sdf, rename(col)) for col in internal.data_spark_column_names
-                ],
-                data_fields=[field.copy(name=rename(field.name)) for field in internal.data_fields],
-            )
-
-        left_internal = self._internal.resolved_copy
-        right_internal = resolve(right._internal, "right")
+        left_internal = _resolve_internal_merge_frame(self._internal)
+        right_internal = _resolve_internal_merge_frame(right._internal, "right")
 
         left_table = left_internal.spark_frame.alias("left_table")
         right_table = right_internal.spark_frame.alias("right_table")
 
-        left_key_columns = [scol_for(left_table, label) for label in left_key_names]
-        right_key_columns = [scol_for(right_table, label) for label in right_key_names]
+        how = validate_how(how)
 
-        join_condition = reduce(
-            lambda x, y: x & y,
-            [lkey == rkey for lkey, rkey in zip(left_key_columns, right_key_columns)],
-        )
```

Comment:
This logic has been moved to `_resolve_merge_join_condition_from_key_names ` in the global scope.
(not including the `validate_how` call)